### PR TITLE
#1000 - add ability to pass function to setProps for prev props

### PR DESCRIPTION
**Pull request by treshugart**
Implements #1000.

### CODE REVIEW

File: src/with-props.js [hunk 1]
```diff
@@ -178,5 +181,8 @@ export function getProps (elem) {
 }
 
 export function setProps (elem, props) {
+  if (typeof props === 'function') {
```
Comment (Hotell) [inline src/with-props.js]: we should have for these cases general facade-lang.js file

`typeof props === 'function'` => `isFunction(props)` etc

**Comment by treshugart** [comment]
@Hotell maybe it'd be good to so the `set*` on the instance now?

**Comment by Hotell** [comment]
hmm so I'm not 100% on board with this implementation/feature

- accessing just props via `this.props` is OK ( the same as react/preact does )
- setting props async via `this.props = {}` or `this.props = updateMe()` feels very strange and confusing. as I've said I will rather preffer API constraints/design which is battle-proofed so rather `setProps(props: object | Function )`
- this may overall add confusion to consumer:
  -- defining props for component is done via static props ( kind of like React does for propTypes )
  -- setting prop value directly via `this.foo="hello"` is sync
  -- setting prop value via `this.prop`.foo="hello" is async
  -- setting via `this.props` is possible only via internal component calls not from outside 
  -- from typed JS view - user has to provide definitions for all props via generics and then again annotate class instance properties

```ts
type Props = {
 test: string
}
class MyComponent extends Component<Props> {
  // Defines the props.
  static props = {
    test: propString
  }

  // inferred as from generic `<Props>` definition ( default props kinda style, but not really )
  props = {
    test: 'some value'
  }
  
  // again I have to define and annotate class prop on instance
  test: string;

}

customElements.define('x-test', MyComponent);

////

const elem = new MyComponent();

// { test: 'some value' }
elem.props = { test: 'another value' };

// 'another value'
elem.test;

// 'another value'
elem.test = 'another value again' ;

// 'another value again'
elem.test;
elem.props.test;
```

**Comment by Hotell** [comment]
Hmm so I was thinking a little today and actually having props on `this.props` is fantastic ( as a author of a component (with typed JS) I dont have to specify those props again on instance, because from my perspective I will always operate with `this.props` and `this.setProps()`. If someone want's to consume the component by manually instantiating it and setting props directly ( he wont get support for autocomplete from type definition ), I'm perfectly fine with that use case.

So yeah I just vote `props` setter to `setProps()` instance method change, otherwise LGTM

**Comment by treshugart** [comment]
I feel having a `setProps()` and a `props` getter is super confusing. In fact, I'm not even sure that having a `setProps` is necessary when you can just `Object.assign()`. We also don't have the same problems as React where prev props might be different at the time of setting state. Even though setting props from any angle is async in Skate, the prop values reflect immediately. I'd be fine having just a `props` getter on the `prototype`, I think.

**Comment by treshugart** [comment]
> setting prop value directly via this.foo="hello" is sync

That's actually also async. The value is reflected immediately, but a render is still queued.

**Comment by treshugart** [comment]
Some changes in this PR:

- Subclass `static props` do not automatically return parent class definitions if the subclass has overridden them. This is because there's no way to undo the merging, for example if you only want a subset, not a superset. It's also more explicit.
- Instance `props` setter (old `props(object)`) only affects declared props. Undeclared props are not set anymore. To do this use `Object.assign(host)`.
- Instance `props` setter is always async, just like normal prop setting.
- Property values *always* propagate immediately rather than in batches, renders only happen in batches. This prevents pitfalls where you might want to use a previous value after setting it to affect a future value. This could be happening in different parts of the application. In React, you'd use `setState(callback)` in order to do this.

**Pull request merged**